Panic related to ChaCha20-based header protection

We're running v0.14.0 (fork with minor changes) in production and we encounter the following panic from time to time.

```
panic: chacha20: counter overflow

goroutine 53078166 [running]:
.../vendor/github.com/marten-seemann/chacha20.(*Cipher).XORKeyStream(0xc0031697d8, 0xc00d4d2dc0, 0x5, 0x5, 0xc00d4d2dc0, 0x5, 0x5)
	/go/src/.../vendor/github.com/marten-seemann/chacha20/chacha_generic.go:167 +0xc3b
.../vendor/github.com/marten-seemann/chacha20.XORKeyStream(0xc00d4d2dc0, 0x5, 0x5, 0xc00d4d2dc0, 0x5, 0x5, 0xc00d4d2de5, 0xc00d4d2dc5)
	/go/src/.../vendor/github.com/marten-seemann/chacha20/chacha_generic.go:237 +0x12d
.../vendor/github.com/Psiphon-Labs/quic-go/internal/handshake.(*chachaHeaderProtector).apply(0xc00d4d2dc0, 0xc00e208009, 0x10, 0x5a3, 0xc00e208000, 0xc00e208005, 0x2, 0x5a7)
	/go/src/.../vendor/github.com/Psiphon-Labs/quic-go/internal/handshake/header_protector.go:109 +0xa9
.../vendor/github.com/Psiphon-Labs/quic-go/internal/handshake.(*chachaHeaderProtector).EncryptHeader(0xc00d4d2dc0, 0xc00e208009, 0x10, 0x5a3, 0xc00e208000, 0xc00e208005, 0x2, 0x5a7)
	/go/src/.../vendor/github.com/Psiphon-Labs/quic-go/internal/handshake/header_protector.go:98 +0x77
.../vendor/github.com/Psiphon-Labs/quic-go/internal/handshake.(*updatableAEAD).EncryptHeader(0xc0054fab40, 0xc00e208009, 0x10, 0x5a3, 0xc00e208000, 0xc00e208005, 0x2, 0x5a7)
	/go/src/.../vendor/github.com/Psiphon-Labs/quic-go/internal/handshake/updatable_aead.go:246 +0x8a
.../vendor/github.com/Psiphon-Labs/quic-go.(*packetPacker).writeAndSealPacketWithPadding(0xc00cd6aa90, 0xc0044ec5a0, 0xc00aa7f5e0, 0x1, 0x1, 0x0, 0x4cd, 0x0, 0x3, 0x7f9a46a5acd8, ...)
	/go/src/.../vendor/github.com/Psiphon-Labs/quic-go/packet_packer.go:579 +0x456
.../vendor/github.com/Psiphon-Labs/quic-go.(*packetPacker).writeAndSealPacket(0xc00cd6aa90, 0xc0044ec5a0, 0xc00aa7f5e0, 0x1, 0x1, 0x0, 0x4cd, 0x3, 0x7f9a46a5acd8, 0xc0054fab40, ...)
	/go/src/.../vendor/github.com/Psiphon-Labs/quic-go/packet_packer.go:534 +0x114
.../vendor/github.com/Psiphon-Labs/quic-go.(*packetPacker).maybePackAppDataPacket(0xc00cd6aa90, 0x0, 0x0, 0x0)
	/go/src/.../vendor/github.com/Psiphon-Labs/quic-go/packet_packer.go:401 +0x26a
.../vendor/github.com/Psiphon-Labs/quic-go.(*packetPacker).PackPacket(0xc00cd6aa90, 0xc003169c00, 0x0, 0xc0058630a0)
	/go/src/.../vendor/github.com/Psiphon-Labs/quic-go/packet_packer.go:274 +0x41
.../vendor/github.com/Psiphon-Labs/quic-go.(*session).sendPacket(0xc0027faa00, 0x1, 0xc003169f18, 0xc011d9eae0)
	/go/src/.../vendor/github.com/Psiphon-Labs/quic-go/session.go:1272 +0x80
.../vendor/github.com/Psiphon-Labs/quic-go.(*session).sendPackets(0xc0027faa00, 0x10a164e8d5836, 0x12cbc40)
	/go/src/.../vendor/github.com/Psiphon-Labs/quic-go/session.go:1187 +0x270
.../vendor/github.com/Psiphon-Labs/quic-go.(*session).run(0xc0027faa00, 0x0, 0x0)
	/go/src/.../vendor/github.com/Psiphon-Labs/quic-go/session.go:541 +0x5af
created by .../vendor/github.com/Psiphon-Labs/quic-go.(*baseServer).createNewSession
	/go/src/.../vendor/github.com/Psiphon-Labs/quic-go/server.go:461 +0x3dd
```

While ChaCha20 allows 2^32 blocks (https://tools.ietf.org/html/rfc7539#section-2.3.2), we observe that the `chacha20` code will panic when the counter is at 2^32-1 and processing just 1 block, even though that should be a valid case: https://github.com/marten-seemann/chacha20/blob/36564989294fee5f3957d3e3fbfc655e10786ec0/chacha_generic.go#L165-L16.

Since QUIC ChaCha20-based header protection (https://tools.ietf.org/html/draft-ietf-quic-tls-24#section-5.4.4) initializes the counter with "4 bytes of [...] sampled ciphertext", it appears that it's possible that a counter value of 2^32-1 (0xffffffff) is selected and then `chacha20` will panic.

We note that the upstream `x/crypto/chacha20` has the same panic bahavior, and a number of other Go ChaCha20 implementations also panic when the counter is 2^32-1 and there's only 1 block to process. But it appears that OpenSSL allows a full 2^32 blocks to be processed before entering its overflow logic: https://github.com/openssl/openssl/blob/706457b7bda7fdbab426b8dce83b318908339da4/crypto/evp/e_chacha20_poly1305.c#L94-L104.

So, we believe that the correct fix is to allow the counter value 2^31-1 and only panic if more than 1 block is processed beyond that point.

We have implemented a proposed fix: https://github.com/Psiphon-Labs/chacha20/commit/899a4be528633ecf678f45e4f6b177d0f89b9e7c.

--- a/go.mod
+++ b/go.mod
@@ -12,7 +12,7 @@ require (
 	github.com/marten-seemann/qtls v0.9.0
 	github.com/onsi/ginkgo v1.11.0
 	github.com/onsi/gomega v1.8.1
-	golang.org/x/crypto v0.0.0-20200221231518-2aa609cf4a9d
+	golang.org/x/crypto v0.0.0-20200423211502-4bdfaf469ed5
 	golang.org/x/net v0.0.0-20190404232315-eb5bcb51f2a3
 	golang.org/x/sync v0.0.0-20190423024810-112230192c58
 )

--- a/internal/handshake/aead_test.go
+++ b/internal/handshake/aead_test.go
@@ -1,6 +1,7 @@
 package handshake
 
 import (
+	"bytes"
 	"crypto/aes"
 	"crypto/cipher"
 	"crypto/rand"
@@ -77,6 +78,24 @@ var _ = Describe("Long Header AEAD", func() {
 						Expect(header).To(Equal([]byte{0xb5, 1, 2, 3, 4, 5, 6, 7, 8, 0xde, 0xad, 0xbe, 0xef}))
 					}
 					Expect(lastFourBitsDifferent).To(BeNumerically(">", 75))
+				})
+
+				It("encrypts and encrypts the header, for a 0xfff..fff sample", func() {
+					sealer, opener := getSealerAndOpener()
+					var lastFourBitsDifferent int
+					for i := 0; i < 100; i++ {
+						sample := bytes.Repeat([]byte{0xff}, 16)
+						header := []byte{0xb5, 1, 2, 3, 4, 5, 6, 7, 8, 0xde, 0xad, 0xbe, 0xef}
+						sealer.EncryptHeader(sample, &header[0], header[9:13])
+						if header[0]&0xf != 0xb5&0xf {
+							lastFourBitsDifferent++
+						}
+						Expect(header[0] & 0xf0).To(Equal(byte(0xb5 & 0xf0)))
+						Expect(header[1:9]).To(Equal([]byte{1, 2, 3, 4, 5, 6, 7, 8}))
+						Expect(header[9:13]).ToNot(Equal([]byte{0xde, 0xad, 0xbe, 0xef}))
+						opener.DecryptHeader(sample, &header[0], header[9:13])
+						Expect(header).To(Equal([]byte{0xb5, 1, 2, 3, 4, 5, 6, 7, 8, 0xde, 0xad, 0xbe, 0xef}))
+					}
 				})
 
 				It("fails to decrypt the header when using a different sample", func() {

--- a/internal/handshake/header_protector.go
+++ b/internal/handshake/header_protector.go
@@ -3,7 +3,6 @@ package handshake
 import (
 	"crypto/aes"
 	"crypto/cipher"
-	"crypto/rand"
 	"encoding/binary"
 	"fmt"
 
@@ -93,27 +92,10 @@ func newChaChaHeaderProtector(suite *qtls.CipherSuiteTLS13, trafficSecret []byte
 }
 
 func (p *chachaHeaderProtector) DecryptHeader(sample []byte, firstByte *byte, hdrBytes []byte) {
-	// Workaround for https://github.com/lucas-clemente/quic-go/issues/2326.
-	// The ChaCha20 implementation panics when the nonce is 0xffffffff.
-	// Don't apply header protection in that case.
-	// The packet will end up undecryptable, but it only applies to 1 in 2^32 packets.
-	if sample[0] == 0xff && sample[1] == 0xff && sample[2] == 0xff && sample[3] == 0xff {
-		return
-	}
 	p.apply(sample, firstByte, hdrBytes)
 }
 
 func (p *chachaHeaderProtector) EncryptHeader(sample []byte, firstByte *byte, hdrBytes []byte) {
-	// Workaround for https://github.com/lucas-clemente/quic-go/issues/2326.
-	// The ChaCha20 implementation panics when the nonce is 0xffffffff.
-	// Apply header protection with a random mask, in order to not leak any data.
-	// The packet will end up undecryptable, but this only applies to 1 in 2^32 packets.
-	if sample[0] == 0xff && sample[1] == 0xff && sample[2] == 0xff && sample[3] == 0xff {
-		if _, err := rand.Read(p.mask[:]); err != nil {
-			panic("couldn't get rand for ChaCha20 bug workaround")
-		}
-		p.applyMask(firstByte, hdrBytes)
-	}
 	p.apply(sample, firstByte, hdrBytes)
 }
 

